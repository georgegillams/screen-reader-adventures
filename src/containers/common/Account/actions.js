import { createRoutine } from 'redux-saga-routines';

import { LOGOUT, REQUEST_VERIFICATION_EMAIL } from './constants';

export const logout = createRoutine(LOGOUT);
export const requestVerificationEmail = createRoutine(REQUEST_VERIFICATION_EMAIL);
