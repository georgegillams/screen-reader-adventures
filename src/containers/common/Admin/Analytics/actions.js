import { createRoutine } from 'redux-saga-routines';

import { LOAD } from './constants';

export const load = createRoutine(LOAD);
