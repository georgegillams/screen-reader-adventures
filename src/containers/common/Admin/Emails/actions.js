import { createRoutine } from 'redux-saga-routines';

import { LOAD, RESEND } from './constants';

export const load = createRoutine(LOAD);
export const resend = createRoutine(RESEND);
