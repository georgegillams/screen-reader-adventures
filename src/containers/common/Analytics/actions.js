import { createRoutine } from 'redux-saga-routines';

import { SEND_ANALYTIC } from './constants';

export const sendAnalytic = createRoutine(SEND_ANALYTIC);
