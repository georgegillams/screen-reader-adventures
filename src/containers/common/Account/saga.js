import { takeLatest, put, call } from 'redux-saga/effects';

import request from 'client-utils/common/request';
import apiStructure from 'helpers/common/apiStructure';

import { logout, requestVerificationEmail } from './actions';
import { setUser } from 'containers/common/Authenticator/actions';

export function* doLogout() {
  const requestURL = apiStructure.logout.fullPath;

  try {
    yield put(logout.request());

    const result = yield call(request, requestURL, {
      method: 'POST',
    });

    if (result.error) {
      yield put(logout.failure(result));
    } else {
      yield put(logout.success(result));
      yield put(setUser.trigger(null));
    }
  } catch (err) {
    yield put(logout.failure(err));
  } finally {
    yield put(logout.fulfill());
  }
}

export function* doRequestVerificationEmail() {
  const requestURL = apiStructure.requestVerificationEmail.fullPath;

  try {
    yield put(requestVerificationEmail.request());

    const result = yield call(request, requestURL, {
      method: 'POST',
    });

    if (result.error) {
      yield put(requestVerificationEmail.failure(result));
    } else {
      yield put(requestVerificationEmail.success(result));
    }
  } catch (err) {
    yield put(requestVerificationEmail.failure(err));
  } finally {
    yield put(requestVerificationEmail.fulfill());
  }
}

export default function* saga() {
  yield takeLatest(logout.TRIGGER, doLogout);
  yield takeLatest(requestVerificationEmail.TRIGGER, doRequestVerificationEmail);
}
