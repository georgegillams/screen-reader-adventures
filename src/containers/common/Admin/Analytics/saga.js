import { takeLatest, put, call } from 'redux-saga/effects';

import request from 'client-utils/common/request';
import apiStructure from 'helpers/common/apiStructure';

import { load } from './actions';

export function* doLoad() {
  const requestURL = apiStructure.loadAnalyticsSummary.fullPath;

  try {
    yield put(load.request());

    const result = yield call(request, requestURL);

    if (result.error) {
      yield put(load.failure(result));
    } else {
      yield put(load.success(result.analytics));
    }
  } catch (err) {
    yield put(load.failure(err));
  } finally {
    yield put(load.fulfill());
  }
}

export default function* saga() {
  yield takeLatest(load.TRIGGER, doLoad);
}
