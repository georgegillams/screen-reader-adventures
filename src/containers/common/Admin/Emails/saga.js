import { takeLatest, put, call, select } from 'redux-saga/effects';

import request from 'client-utils/common/request';
import apiStructure from 'helpers/common/apiStructure';

import { load, resend } from './actions';
import { selectState } from './selectors';

export function* doLoad() {
  const requestURL = apiStructure.loadEmails.fullPath;

  try {
    yield put(load.request());

    const result = yield call(request, requestURL);

    if (result.error) {
      yield put(load.failure(result));
    } else {
      yield put(load.success(result.emails));
    }
  } catch (err) {
    yield put(load.failure(err));
  } finally {
    yield put(load.fulfill());
  }
}

export function* doResend() {
  const currentState = yield select(selectState());
  const { emailToResend } = currentState;
  const requestURL = apiStructure.resendEmail.fullPath;

  try {
    yield put(resend.request());

    const result = yield call(request, requestURL, {
      method: 'POST',
      body: JSON.stringify(emailToResend),
      headers: {
        'Content-Type': 'application/json',
      },
    });

    if (result.error) {
      yield put(resend.failure(result));
    } else {
      yield put(resend.success(result.emails));
      yield put(load.trigger());
    }
  } catch (err) {
    yield put(resend.failure(err));
  } finally {
    yield put(resend.fulfill());
  }
}

export default function* saga() {
  yield takeLatest(load.TRIGGER, doLoad);
  yield takeLatest(resend.TRIGGER, doResend);
}
