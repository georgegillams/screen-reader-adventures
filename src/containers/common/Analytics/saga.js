import { takeLatest, put, call, select } from 'redux-saga/effects';

import request from 'client-utils/common/request';
import apiStructure from 'helpers/common/apiStructure';

import { selectState } from './selectors';
import { sendAnalytic } from './actions';

export function* doSendAnalytic() {
  const currentState = yield select(selectState());
  const { analytic } = currentState;
  const requestURL = apiStructure.createAnalytic.fullPath;

  try {
    yield put(sendAnalytic.request());

    const result = yield call(request, requestURL, {
      method: 'POST',
      body: JSON.stringify(analytic),
      headers: {
        'Content-Type': 'application/json',
      },
    });

    if (result.error) {
      yield put(sendAnalytic.failure(result));
    } else {
      yield put(sendAnalytic.success(result));
    }
  } catch (err) {
    yield put(sendAnalytic.failure(err));
  } finally {
    yield put(sendAnalytic.fulfill());
  }
}

export default function* saga() {
  yield takeLatest(sendAnalytic.TRIGGER, doSendAnalytic);
}
