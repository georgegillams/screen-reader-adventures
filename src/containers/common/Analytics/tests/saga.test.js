import { put, takeLatest } from 'redux-saga/effects';

import { sendAnalytic } from '../actions';

import saga, { doSendAnalytic } from '../saga';

describe('Analytics saga', () => {
  let mainSaga;

  beforeEach(() => {
    mainSaga = saga();
  });

  it('Should send analytic on TRIGGER', () => {
    const takeLatestDescriptor = mainSaga.next().value;
    expect(takeLatestDescriptor).toEqual(takeLatest(sendAnalytic.TRIGGER, doSendAnalytic));
  });

  describe('sendAnalytic actions', () => {
    let sendAnalyticGenerator;

    const response = {
      browser: 'Firefox',
      os: 'macOS',
      status: 200,
    };

    beforeEach(() => {
      sendAnalyticGenerator = doSendAnalytic();

      const selectDescriptor = sendAnalyticGenerator.next().value;
      expect(selectDescriptor).toMatchSnapshot();
    });

    it('Should call sendAnalytic.success on successful API call', () => {
      sendAnalyticGenerator.next({ analytic: { browser: 'Firefox' } });
      sendAnalyticGenerator.next();
      const putSuccess = sendAnalyticGenerator.next(response).value;
      sendAnalyticGenerator.next();

      expect(putSuccess).toEqual(put(sendAnalytic.success(response)));
    });

    it('Should call sendAnalytic.failure if an error is returned', () => {
      sendAnalyticGenerator.next({ analytic: { browser: 'Firefox' } });
      sendAnalyticGenerator.next();
      const response = { error: 'auth', errorMessage: 'some error' };
      const putFailure = sendAnalyticGenerator.next(response).value;

      expect(putFailure).toEqual(put(sendAnalytic.failure(response)));
    });

    it('Should call sendAnalytic.failure if an exception occurs', () => {
      sendAnalyticGenerator.next({ analytic: { browser: 'Firefox' } });
      const response = new Error('Some error');
      const putFailure = sendAnalyticGenerator.throw(response).value;

      expect(putFailure).toEqual(put(sendAnalytic.failure(response)));
    });
  });
});
