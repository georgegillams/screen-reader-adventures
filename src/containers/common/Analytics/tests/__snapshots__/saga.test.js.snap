// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Analytics saga sendAnalytic actions Should call sendAnalytic.failure if an error is returned 1`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [],
    "selector": [Function],
  },
  "type": "SELECT",
}
`;

exports[`Analytics saga sendAnalytic actions Should call sendAnalytic.failure if an exception occurs 1`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [],
    "selector": [Function],
  },
  "type": "SELECT",
}
`;

exports[`Analytics saga sendAnalytic actions Should call sendAnalytic.success on successful API call 1`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [],
    "selector": [Function],
  },
  "type": "SELECT",
}
`;
